---
title: "LATER"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{LATER}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
```

```{r setup}
library(LATER)
```

# Load Drosphila reference annotation 
```{r}
# load reference annotation 
annot_path <- system.file("exdata/dm6.annot.gtf.gz", package="LATER")
ref_annot <- rtracklayer::import.gff(annot_path)
```


# Create links database 

Links database represents the isoforms links between 5'-3' 

```{r, message=FALSE, warning=FALSE}
database <- prepareLinksDatabase(ref_annot, 
                                 tss.window = 50, 
                                 tes.window = 150) 
```

The database classifies isoform and genes in the context of their 5'/3' regulation

```{r, message=FALSE, warning=FALSE}
head(database$pairDataBase,3)
```

# Count 5'-3' links 

Using the database we can quantify alignments generated using `minimap` to assign reads to TSS and PA sites. Results are normalized in counts per million.

```{r,  message=FALSE, warning=FALSE}
bamPath <- system.file("exdata/testBam.bam", package = 'LATER')
countsLongReads  <- countLinks(bamPath, database) 
head(countsLongReads) 
```

# Calculate promoter dominance 

```{r,  message=FALSE, warning=FALSE}
promoterDominance <- calculatePromoterDominance(countsLongReads, database$pairDataBase)
```

```{r, include=FALSE}
head(promoterDominance,3)
```


# Transcriptional bias 

Transcriptional bias... in here we use fisher.test 

```{r,  message=FALSE, warning=FALSE}
transcriptionalBias <-
  estimateTranscriptionalBias(promoterDominance, method = "fisher")
head(transcriptionalBias)
```






