---
title: "Long-reads-based Alternative Termination Estimation and Recognition (LATER)"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{LATER}
  %\VignetteEncoding{UTF-8}
  %\VignetteEngine{knitr::rmarkdown}
editor_options: 
  chunk_output_type: inline
---

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
```


```{r, eval = FALSE}
library(devtools)
devtools::install_github("hilgers-lab/LATER", build = TRUE)
```


# Reference annotation

```{r, eval = FALSE, message=FALSE, warning=FALSE}
# Download ensembl annotation 
Annotation_url <- "https://ftp.ensembl.org/pub/release-109/gtf/drosophila_melanogaster/Drosophila_melanogaster.BDGP6.32.109.gtf.gz"
annotation_file_name <- "dmel_reference_annotation.gtf.gz"
download.file(Annotation_url, annotation_file_name)
# read file 
reference_annotation <- rtracklayer::import.gff("dmel_reference_annotation.gtf.gz")
protein_coding_exons <- reference_annotation[reference_annotation$type == "exon" & reference_annotation$gene_biotype == "protein_coding"]
protein_coding_genes <- reference_annotation[reference_annotation$type == "gene" & reference_annotation$gene_biotype == "protein_coding"]
```


```{r, include = FALSE, message=FALSE, warning=FALSE}
require(dplyr)
ref_path <- system.file("exdata/dmel_reference_annotation.gtf.gz", 
                        package = 'LATER')
reference_annotation <- rtracklayer::import.gff(ref_path)
protein_coding_exons <- reference_annotation[reference_annotation$type == "exon" & reference_annotation$gene_biotype == "protein_coding"]
protein_coding_genes <- reference_annotation[reference_annotation$type == "gene" & reference_annotation$gene_biotype == "protein_coding"]
```


## Create IsoformAnnotation object 

```{r}
library(LATER)
isoformData <- prepareIsoformDatabase(protein_coding_exons, 
                                        tss.window = 50, 
                                        tes.window = 150)
```

IsoformAnnotation contains the following slots:
```{r}
slotNames(isoformData)
```

This allows to retrieve data easily using: 
```{r}
showGene(isoformData, "FBgn0067779")
```

### Using custom TSS or 3'end databases 

To use custom TSS or 3'end database to increase the detection of 5'/3' ends in the long read data the following functions can be used. 

```{r, eval=FALSE}
url <- "https://ftp.ensembl.org/pub/release-109/gtf/drosophila_melanogaster/Drosophila_melanogaster.BDGP6.32.109.gtf.gz"
file_name <- " dmel_tss_annotation.gtf.gz "
download.file(url, file_name)
ref_tss_annot <- rtracklayer::import.gff("dmel_tss_annotation.gtf.gz")
```

This database can be used to increase the TSS repertoire of the IsoformDatabase 

```{r, eval=FALSE}
isoformData  <- addPromoterDatabase(isoformData, 
                                         ref_tss_annot, 
                                         protein_coding_genes, 
                                         window = 50)
```

# Counting 5'-3' isoforms 

countData object is a LATER object containing all the relevant information for full-length quantification. The slots added at this step are: `LATER@readAssignments` that contains the read_id and their assigned features, `LATER@isoformCounts` that contains the summarized counts per 5'-3' isoform. 

```{r}
bamPath <- system.file("exdata/testBam.bam", package = 'LATER')
countData <- countLinks(bamPath, isoformData)
```

Read ids can be easily exported using the follwing code: 

```{r, eval=FALSE}
readr::write_tsv(readAssignments(countData), "read_assignments.txt")
```

bam files can be easily substeed using this read ids using the following code on linux

```{bash}
samtools view -N read_assignments.txt -o filtered_output.bam output.bam
```

# Statistical testing of 5’-3’ couplings. 

Using this counts we can now estimate 5’-3’ links from the read counts, using multinomial testing. 

```{r}
gene_bias_estimates <- estimatePromoterDominance(countData, isoformData, method="fisher")
```







